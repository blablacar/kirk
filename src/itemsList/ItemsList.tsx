import React, { FunctionComponent } from 'react'
import cc from 'classcat'

import { A11yProps, pickA11yProps } from '../_utils/interfaces'
import { Divider } from '../divider'
import { ItemCheckboxProps } from '../itemCheckbox/ItemCheckbox'
import { ItemChoiceProps } from '../itemChoice'
import { ItemRadioProps } from '../itemRadio/ItemRadio'
import { StyledItemsList } from './ItemsList.style'

export const ItemsListDivider: FunctionComponent = () => null

// NOTE: react-hot-loader will update dynamically the type and break type comparisons.
// A pre-rendered type need to be used to fix it.
// See: https://github.com/gaearon/react-hot-loader#checking-element-types
const ItemsListDividerType = (<ItemsListDivider />).type

export type ItemsListChild =
  | React.ReactElement<ItemChoiceProps>
  | React.ReactElement<ItemRadioProps>
  | React.ReactElement<ItemCheckboxProps>
  | null

export type ItemsListProps = A11yProps &
  Readonly<{
    children: ItemsListChild[]
    withSeparators?: boolean
    className?: string
    keyGenerator?: (index: number) => string | number
  }>

export const ItemsList = (props: ItemsListProps) => {
  const {
    children,
    className = '',
    withSeparators = false,
    keyGenerator = (i: number) => i,
  } = props
  const a11yAttrs = pickA11yProps<ItemsListProps>(props)

  return (
    <StyledItemsList className={cc(['kirk-items-list', className])} {...a11yAttrs}>
      {children.map((item, index) => {
        if (item.type === ItemsListDividerType || item.type === undefined) {
          return null
        }

        const isLast = children.length === index + 1
        const hasSeparator =
          !isLast && (children[index + 1].type === ItemsListDividerType || withSeparators)

        return (
          <li className={cc(['kirk-items-list-item'])} key={keyGenerator(index)}>
            {item}
            {hasSeparator && <Divider />}
          </li>
        )
      })}
    </StyledItemsList>
  )
}
