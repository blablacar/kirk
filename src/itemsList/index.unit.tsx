import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'

import ItemChoice from 'itemChoice'
import ItemsList from './index'

describe('ItemsList', () => {
  it('Should render an unordered list', () => {
    const itemChoice = renderer
      .create(
        <ItemsList>
          <ItemChoice label="Item 1" href="#1" />
          <ItemChoice label="Item 2" href="#2" />
          <ItemChoice label="Item 3" href="#3" />
        </ItemsList>,
      )
      .toJSON()
    expect(itemChoice).toMatchSnapshot()
  })
  it('Should render an unordered list with the separators classname', () => {
    const itemChoice = renderer
      .create(
        <ItemsList withSeparators>
          <ItemChoice label="Item 1" href="#1" />
          <ItemChoice label="Item 2" href="#2" />
          <ItemChoice label="Item 3" href="#3" />
        </ItemsList>,
      )
      .toJSON()
    expect(itemChoice).toMatchSnapshot()
  })
  it('Should render an unordered list with some separators', () => {
    const itemChoice = renderer
      .create(
        <ItemsList>
          <ItemChoice label="Item 1" href="#1" />
          <ItemChoice label="Item 2" href="#2" />
          <ItemsList.Divider />
          <ItemChoice label="Item 3" href="#3" />
        </ItemsList>,
      )
      .toJSON()
    expect(itemChoice).toMatchSnapshot()
  })
})
