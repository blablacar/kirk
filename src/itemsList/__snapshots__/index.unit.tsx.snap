// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsList Should render an unordered list 1`] = `
.c0 {
  margin: 0;
  font-weight: 400;
}

.c0.kirk-text-button {
  color: #054752;
  font-size: 16px;
  line-height: 20px;
}

.c0.kirk-text-body,
.c0.kirk-text-bodyStrong {
  color: #708C91;
  font-size: 16px;
  line-height: 20px;
}

.c0.kirk-text-caption {
  color: #708C91;
  font-size: 13px;
  line-height: 16px;
}

.c0.kirk-text-title,
.c0.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c0.kirk-text-display1 {
  color: #054752;
  font-size: 82px;
  line-height: 82px;
}

.c0.kirk-text-display2 {
  color: #054752;
  font-size: 30px;
  line-height: 1.06;
}

.c0.kirk-text-subheader,
.c0.kirk-text-subheaderStrong {
  color: #054752;
  font-size: 22px;
  line-height: 24px;
}

.c0.kirk-text-title,
.c0.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c0.kirk-text-bodyStrong,
.c0.kirk-text-subheaderStrong,
.c0.kirk-text-titleStrong {
  font-weight: 500;
}

.c1 {
  display: inline-block;
  position: relative;
}

.c1 .kirk-icon-badge {
  position: absolute;
  bottom: 12px;
  left: 12px;
}

<ul
  className="jsx-1825249063 kirk-items-list"
  role=""
>
  <li
    className="jsx-1825249063 kirk-items-list-item"
  >
    <a
      className="jsx-1542968643 kirk-item kirk-item--clickable kirk-item-choice"
      href="#1"
    >
      <div
        className="jsx-1542968643 kirk-item-leftText"
      >
        <div
          className="kirk-text kirk-text-title c0"
          style={
            Object {
              "color": "#054752",
            }
          }
        >
          Item 1
        </div>
        
      </div>
      
      <div
        className="jsx-1542968643 kirk-item-rightAddon"
      >
        <svg
          aria-hidden={true}
          className="kirk-icon c1"
          fill="#708C91"
          height={24}
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <polyline
            fill="none"
            points="9 18 15 12 9 6"
            stroke="#708C91"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeMiterlimit="10"
            strokeWidth="2"
          />
        </svg>
      </div>
    </a>
  </li>
  <li
    className="jsx-1825249063 kirk-items-list-item"
  >
    <a
      className="jsx-1542968643 kirk-item kirk-item--clickable kirk-item-choice"
      href="#2"
    >
      <div
        className="jsx-1542968643 kirk-item-leftText"
      >
        <div
          className="kirk-text kirk-text-title c0"
          style={
            Object {
              "color": "#054752",
            }
          }
        >
          Item 2
        </div>
        
      </div>
      
      <div
        className="jsx-1542968643 kirk-item-rightAddon"
      >
        <svg
          aria-hidden={true}
          className="kirk-icon c1"
          fill="#708C91"
          height={24}
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <polyline
            fill="none"
            points="9 18 15 12 9 6"
            stroke="#708C91"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeMiterlimit="10"
            strokeWidth="2"
          />
        </svg>
      </div>
    </a>
  </li>
  <li
    className="jsx-1825249063 kirk-items-list-item"
  >
    <a
      className="jsx-1542968643 kirk-item kirk-item--clickable kirk-item-choice"
      href="#3"
    >
      <div
        className="jsx-1542968643 kirk-item-leftText"
      >
        <div
          className="kirk-text kirk-text-title c0"
          style={
            Object {
              "color": "#054752",
            }
          }
        >
          Item 3
        </div>
        
      </div>
      
      <div
        className="jsx-1542968643 kirk-item-rightAddon"
      >
        <svg
          aria-hidden={true}
          className="kirk-icon c1"
          fill="#708C91"
          height={24}
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <polyline
            fill="none"
            points="9 18 15 12 9 6"
            stroke="#708C91"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeMiterlimit="10"
            strokeWidth="2"
          />
        </svg>
      </div>
    </a>
  </li>
</ul>
`;

exports[`ItemsList Should render an unordered list with some separators 1`] = `
.c0 {
  margin: 0;
  font-weight: 400;
}

.c0.kirk-text-button {
  color: #054752;
  font-size: 16px;
  line-height: 20px;
}

.c0.kirk-text-body,
.c0.kirk-text-bodyStrong {
  color: #708C91;
  font-size: 16px;
  line-height: 20px;
}

.c0.kirk-text-caption {
  color: #708C91;
  font-size: 13px;
  line-height: 16px;
}

.c0.kirk-text-title,
.c0.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c0.kirk-text-display1 {
  color: #054752;
  font-size: 82px;
  line-height: 82px;
}

.c0.kirk-text-display2 {
  color: #054752;
  font-size: 30px;
  line-height: 1.06;
}

.c0.kirk-text-subheader,
.c0.kirk-text-subheaderStrong {
  color: #054752;
  font-size: 22px;
  line-height: 24px;
}

.c0.kirk-text-title,
.c0.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c0.kirk-text-bodyStrong,
.c0.kirk-text-subheaderStrong,
.c0.kirk-text-titleStrong {
  font-weight: 500;
}

.c1 {
  display: inline-block;
  position: relative;
}

.c1 .kirk-icon-badge {
  position: absolute;
  bottom: 12px;
  left: 12px;
}

<ul
  className="jsx-1825249063 kirk-items-list"
  role=""
>
  <li
    className="jsx-1825249063 kirk-items-list-item"
  >
    <a
      className="jsx-1542968643 kirk-item kirk-item--clickable kirk-item-choice"
      href="#1"
    >
      <div
        className="jsx-1542968643 kirk-item-leftText"
      >
        <div
          className="kirk-text kirk-text-title c0"
          style={
            Object {
              "color": "#054752",
            }
          }
        >
          Item 1
        </div>
        
      </div>
      
      <div
        className="jsx-1542968643 kirk-item-rightAddon"
      >
        <svg
          aria-hidden={true}
          className="kirk-icon c1"
          fill="#708C91"
          height={24}
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <polyline
            fill="none"
            points="9 18 15 12 9 6"
            stroke="#708C91"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeMiterlimit="10"
            strokeWidth="2"
          />
        </svg>
      </div>
    </a>
  </li>
  <li
    className="jsx-1825249063 kirk-items-list-item"
  >
    <a
      className="jsx-1542968643 kirk-item kirk-item--clickable kirk-item-choice"
      href="#2"
    >
      <div
        className="jsx-1542968643 kirk-item-leftText"
      >
        <div
          className="kirk-text kirk-text-title c0"
          style={
            Object {
              "color": "#054752",
            }
          }
        >
          Item 2
        </div>
        
      </div>
      
      <div
        className="jsx-1542968643 kirk-item-rightAddon"
      >
        <svg
          aria-hidden={true}
          className="kirk-icon c1"
          fill="#708C91"
          height={24}
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <polyline
            fill="none"
            points="9 18 15 12 9 6"
            stroke="#708C91"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeMiterlimit="10"
            strokeWidth="2"
          />
        </svg>
      </div>
    </a>
  </li>
  <li
    className="jsx-1825249063 kirk-items-list-item kirk-items-list-item--withSeparator"
  >
    <a
      className="jsx-1542968643 kirk-item kirk-item--clickable kirk-item-choice"
      href="#3"
    >
      <div
        className="jsx-1542968643 kirk-item-leftText"
      >
        <div
          className="kirk-text kirk-text-title c0"
          style={
            Object {
              "color": "#054752",
            }
          }
        >
          Item 3
        </div>
        
      </div>
      
      <div
        className="jsx-1542968643 kirk-item-rightAddon"
      >
        <svg
          aria-hidden={true}
          className="kirk-icon c1"
          fill="#708C91"
          height={24}
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <polyline
            fill="none"
            points="9 18 15 12 9 6"
            stroke="#708C91"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeMiterlimit="10"
            strokeWidth="2"
          />
        </svg>
      </div>
    </a>
  </li>
</ul>
`;

exports[`ItemsList Should render an unordered list with the separators classname 1`] = `
.c0 {
  margin: 0;
  font-weight: 400;
}

.c0.kirk-text-button {
  color: #054752;
  font-size: 16px;
  line-height: 20px;
}

.c0.kirk-text-body,
.c0.kirk-text-bodyStrong {
  color: #708C91;
  font-size: 16px;
  line-height: 20px;
}

.c0.kirk-text-caption {
  color: #708C91;
  font-size: 13px;
  line-height: 16px;
}

.c0.kirk-text-title,
.c0.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c0.kirk-text-display1 {
  color: #054752;
  font-size: 82px;
  line-height: 82px;
}

.c0.kirk-text-display2 {
  color: #054752;
  font-size: 30px;
  line-height: 1.06;
}

.c0.kirk-text-subheader,
.c0.kirk-text-subheaderStrong {
  color: #054752;
  font-size: 22px;
  line-height: 24px;
}

.c0.kirk-text-title,
.c0.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c0.kirk-text-bodyStrong,
.c0.kirk-text-subheaderStrong,
.c0.kirk-text-titleStrong {
  font-weight: 500;
}

.c1 {
  display: inline-block;
  position: relative;
}

.c1 .kirk-icon-badge {
  position: absolute;
  bottom: 12px;
  left: 12px;
}

<ul
  className="jsx-1825249063 kirk-items-list kirk-items-list--withSeparators"
  role=""
>
  <li
    className="jsx-1825249063 kirk-items-list-item"
  >
    <a
      className="jsx-1542968643 kirk-item kirk-item--clickable kirk-item-choice"
      href="#1"
    >
      <div
        className="jsx-1542968643 kirk-item-leftText"
      >
        <div
          className="kirk-text kirk-text-title c0"
          style={
            Object {
              "color": "#054752",
            }
          }
        >
          Item 1
        </div>
        
      </div>
      
      <div
        className="jsx-1542968643 kirk-item-rightAddon"
      >
        <svg
          aria-hidden={true}
          className="kirk-icon c1"
          fill="#708C91"
          height={24}
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <polyline
            fill="none"
            points="9 18 15 12 9 6"
            stroke="#708C91"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeMiterlimit="10"
            strokeWidth="2"
          />
        </svg>
      </div>
    </a>
  </li>
  <li
    className="jsx-1825249063 kirk-items-list-item"
  >
    <a
      className="jsx-1542968643 kirk-item kirk-item--clickable kirk-item-choice"
      href="#2"
    >
      <div
        className="jsx-1542968643 kirk-item-leftText"
      >
        <div
          className="kirk-text kirk-text-title c0"
          style={
            Object {
              "color": "#054752",
            }
          }
        >
          Item 2
        </div>
        
      </div>
      
      <div
        className="jsx-1542968643 kirk-item-rightAddon"
      >
        <svg
          aria-hidden={true}
          className="kirk-icon c1"
          fill="#708C91"
          height={24}
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <polyline
            fill="none"
            points="9 18 15 12 9 6"
            stroke="#708C91"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeMiterlimit="10"
            strokeWidth="2"
          />
        </svg>
      </div>
    </a>
  </li>
  <li
    className="jsx-1825249063 kirk-items-list-item"
  >
    <a
      className="jsx-1542968643 kirk-item kirk-item--clickable kirk-item-choice"
      href="#3"
    >
      <div
        className="jsx-1542968643 kirk-item-leftText"
      >
        <div
          className="kirk-text kirk-text-title c0"
          style={
            Object {
              "color": "#054752",
            }
          }
        >
          Item 3
        </div>
        
      </div>
      
      <div
        className="jsx-1542968643 kirk-item-rightAddon"
      >
        <svg
          aria-hidden={true}
          className="kirk-icon c1"
          fill="#708C91"
          height={24}
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <polyline
            fill="none"
            points="9 18 15 12 9 6"
            stroke="#708C91"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeMiterlimit="10"
            strokeWidth="2"
          />
        </svg>
      </div>
    </a>
  </li>
</ul>
`;
